File: src/main/kotlin/school/charset/app/domain/progress/ModuleProgress.kt
package school.charset.app.domain.progress

import school.charset.app.domain.exercise.ExerciseModule
import kotlin.time.Instant

data class ModuleProgress(
    val userId: Long,
    val module: ExerciseModule,
    val level: Int,
    val streak: Int,
    val attempts: Int,
    val errors: Int,
    val lastPlayedAt: Instant?,
) {
    // Auto-advances the level once the streak threshold is hit (while still
    // below the module's max level), then resets the streak so the user has
    // to earn the next bump as well. The level is no longer user-selectable
    // from the UI - the back drives progression entirely.
    fun recordCompletion(correct: Boolean, now: Instant): ModuleProgress {
        val updated = copy(
            attempts = attempts + 1,
            errors = if (correct) errors else errors + 1,
            streak = if (correct) streak + 1 else 0,
            lastPlayedAt = now,
        )
        return if (updated.streak >= STREAK_FOR_LEVEL_UP && updated.level < module.maxLevel) {
            updated.copy(level = updated.level + 1, streak = 0)
        } else {
            updated
        }
    }

    companion object {
        const val STREAK_FOR_LEVEL_UP: Int = 5

        fun initial(userId: Long, module: ExerciseModule): ModuleProgress = ModuleProgress(userId, module, level = 1, streak = 0, attempts = 0, errors = 0, lastPlayedAt = null)
    }
}

File: lib/charset/progress/module_progress.ex
defmodule Charset.Progress.ModuleProgress do
  @moduledoc """
  Per-visitor per-module progression. Pure domain struct - the Ecto mapping
  lives in `Charset.Progress`.
  """

  alias Charset.Exercise.ExerciseModule

  @streak_for_level_up 5

  @enforce_keys [:token, :module, :level, :streak, :attempts, :errors]
  defstruct [:token, :module, :level, :streak, :attempts, :errors, :last_played_at]

  @type t :: %__MODULE__{
          token: String.t(),
          module: ExerciseModule.t(),
          level: pos_integer(),
          streak: non_neg_integer(),
          attempts: non_neg_integer(),
          errors: non_neg_integer(),
          last_played_at: DateTime.t() | nil
        }

  def streak_for_level_up, do: @streak_for_level_up

  @spec initial(String.t(), ExerciseModule.t()) :: t()
  def initial(token, module) do
    %__MODULE__{token: token, module: module, level: 1, streak: 0, attempts: 0, errors: 0}
  end

  @doc """
  Records a completed exercise. Auto-advances the level once the streak
  threshold is hit (while still below the module's max level), then resets
  the streak so the user has to earn the next bump as well. The level is not
  user-selectable - the backend drives progression entirely.
  """
  @spec record_completion(t(), boolean(), DateTime.t()) :: t()
  def record_completion(%__MODULE__{} = progress, correct, now) do
    updated = %{
      progress
      | attempts: progress.attempts + 1,
        errors: if(correct, do: progress.errors, else: progress.errors + 1),
        streak: if(correct, do: progress.streak + 1, else: 0),
        last_played_at: now
    }

    if updated.streak >= @streak_for_level_up and
         updated.level < ExerciseModule.max_level(progress.module) do
      %{updated | level: updated.level + 1, streak: 0}
    else
      updated
    end
  end
end
